### ISSUE TITLE
Illegal offset type when fieldList is passed to Marshaller::merge

### ISSUE THREAD

**Issue by aydot** [cakephp/cakephp#6387]
```
<?php

namespace TestCase;


use Cake\ORM\Entity;
use Cake\ORM\Marshaller;
use Cake\ORM\TableRegistry;
use Cake\TestSuite\TestCase;

class MarshalTest extends TestCase
{
    public function testMergeWithSingleAssociationAndFieldLists()
    {
        $user = new Entity([
                               'first_name' => 'ayo',
                               'last_name' => 'akins'
                           ]);

        $entity = new Entity([
                                 'club_identifier' => '4567',
                                 'user' => $user
                             ]);
        $user->accessible('*', true);
        $entity->accessible('*', true);

        $data = [
            'club_name' => 'Chelsea',
            'user' => [
                'last_name' => 'dee'
            ]
        ];

        $marshall = new Marshaller(TableRegistry::get('ClubMembers'));
        $marshall->merge($entity,
                         $data,
                         [
                             'fieldList' => ['club_identifier', 'user'],
                             'associated' => [
                                 'Users' => [
                                 ]
                             ]
                         ]
        );

        $this->assertSame($user, $entity->user);
    }
}
```
throws
```
Warning Error: Illegal offset type in [../vendor/cakephp/cakephp/src/Datasource/EntityTrait.php, line 575]

Warning Error: Illegal offset type in unset in [../vendor/cakephp/cakephp/src/Datasource/EntityTrait.php, line 576]
```

swapping $assoc passed on line [http://api.cakephp.org/3.0/source-class-Cake.ORM.Marshaller.html#442] with $field fixes it.

**Comment by markstory** [comment]
Thanks for the test case :+1:

**Comment by markstory** [comment]
Fixed with #6391
